### PR TITLE
Fix ability to delete other device from the sync chain

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
Basically this PR reverts https://chromium-review.googlesource.com/c/chromium/src/+/3001364 and updates the tests to be aware in future if Chromium will change the behavior  near this place. 

I had to use patch for `src/components/sync_device_info/device_info_sync_bridge.cc`.

Resolves https://github.com/brave/brave-browser/issues/19561

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
Use one from https://github.com/brave/brave-browser/issues/19561

### CODE REVIEW

File: patches/components-sync_device_info-device_info_sync_bridge.cc.patch [hunk 1]
```diff
@@ -10,7 +10,15 @@ index 69dbe2887fae63f3635423ea9a22c3f6ed5f7248..a9d224f072be5a5da78881300031e4ea
    return specifics;
  }
  
-@@ -439,7 +440,7 @@ absl::optional<ModelError> DeviceInfoSyncBridge::ApplySyncChanges(
+@@ -430,6 +431,7 @@ absl::optional<ModelError> DeviceInfoSyncBridge::ApplySyncChanges(
+     const std::string guid = change->storage_key();
+ 
+     // Reupload local device if it was deleted from the server.
++    if (bool false_as_brave_needs_skip_next_if = false)
```
Comment (mkarolin) [inline patches/components-sync_device_info-device_info_sync_bridge.cc.patch:17]: Seems like we can follow our patching patterns here and use `#define BRAVE_DEVICE_INFO_SYNC_BRIDGE_APPLY_SYNC_CHANGES_SKIP_NEXT_IF if (false)` in the `chromium_src` override.

Comment (darkdh) [reply patches/components-sync_device_info-device_info_sync_bridge.cc.patch:17]: and you can also add comment in the overridden file

Comment (AlexeyBarabash) [reply patches/components-sync_device_info-device_info_sync_bridge.cc.patch:17]: thanks @mkarolin @darkdh , I missed that pattern, I'll fix

Comment (AlexeyBarabash) [reply patches/components-sync_device_info-device_info_sync_bridge.cc.patch]: fixed at https://github.com/brave/brave-core/compare/3bb43a4b5f4bdae7a20fbb13789578af33235bad..2ac8203b714bd992f48571119cb9b8381c59136d

File: chromium_src/components/sync_device_info/device_info_sync_bridge.cc [hunk 2]
```diff
@@ -13,8 +13,14 @@
 #define RefreshLocalDeviceInfoIfNeeded \
   RefreshLocalDeviceInfoIfNeeded_ChromiumImpl
 
+// This macro disables Chromium's block which detects whether the local
+// device record should be re-uppload to server. We disable it because it
```
Comment (mkarolin) [inline chromium_src/components/sync_device_info/device_info_sync_bridge.cc:17]: Typo: re-uploaded

**Review by mkarolin — APPROVED** [review]
Haven't looked at the tests logic, but `chromium_src` and `patches` changes LGTM.

File: chromium_src/components/sync_device_info/device_info_sync_bridge.cc (hunk 2 above)
Comment (AlexeyBarabash) [reply chromium_src/components/sync_device_info/device_info_sync_bridge.cc]: thanks, fixed at 
https://github.com/brave/brave-core/compare/2ac8203b714bd992f48571119cb9b8381c59136d..a2bc0c2cc4bc5b4d89394306abc9d79812fbccd0

**Review by darkdh — APPROVED** [review]

**Pull request merged**